extensions/crypto/01_crypto.js

```diff
@@ -41,10 +42,102 @@
     return arrayBufferView;
   }
 
+  const supportedAlgorithms = {
+    "digest": {
+      "SHA-1": {},
+      "SHA-256": {},
+      "SHA-384": {},
+      "SHA-512": {},
+    },
+  };
+
+  function normalizeAlgorithm(algorithm, op) {
+    if (typeof algorithm == "string") {
+      return normalizeAlgorithm({ name: algorithm }, op);
+    }
+
+    const initialAlgorithm = webidl.converters["Algorithm"](algorithm, {
+      context: "Argument 1",
+    });
+
+    const registeredAlgorithms = supportedAlgorithms[op];
+    const algorithmName = Object.keys(registeredAlgorithms)
+      .find((key) => key.toLowerCase() == initialAlgorithm.name.toLowerCase());
+
+    if (algorithmName === undefined) {
+      throw new DOMException(
+        "Unrecognized algorithm name",
+        "NotSupportedError",
+      );
+    }
+
+    // TODO(caspervonb) Step 6 (create from webidl definition), when the need arises.
+    // See https://www.w3.org/TR/WebCryptoAPI/#dfn-normalize-an-algorithm
+    const normalizedAlgorithm = {};
+    normalizedAlgorithm.name = algorithmName;
+
+    // TODO(caspervonb) Step 9 and 10, when the need arises.
+    // See https://www.w3.org/TR/WebCryptoAPI/#dfn-normalize-an-algorithm
+    return normalizedAlgorithm;
+  }
+
+  // Should match op_crypto_subtle_digest() in extensions/crypto/lib.rs
+  function digestToId(name) {
+    switch (name) {
+      case "SHA-1":
+        return 0;
+      case "SHA-256":
+        return 1;
+      case "SHA-384":
+        return 2;
+      case "SHA-512":
+        return 3;
+    }
+  }
+
+  class SubtleCrypto {
+    constructor() {
+      webidl.illegalConstructor();
+    }
+
+    async digest(algorithm, data) {
+      webidl.assertBranded(this, SubtleCrypto);
+      webidl.requiredArguments(arguments.length, 2);
+
+      algorithm = webidl.converters.AlgorithmIdentifier(algorithm, {
+        context: "Argument 1",
+      });
+
+      data = webidl.converters.BufferSource(data, {
+        context: "Argument 2",
+      });
+
+      data = data.slice(0);
```

Comment:
Data returned from BufferSource can either be a buffer view, or a buffer. You need to normalize it all into a uint8array (taking into account byteOffset and byteLength). If there are no WPT tests for this, I would be happy to write some.

Reply:
> You need to normalize it all into a uint8array
 
Was sure I confirmed that serde_v8 would convert this, but on a second test, indeed it only takes TypedArrays.